src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -33,107 +34,156 @@
         produces = "application/json")
     public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
 
-        int i = 0;
+        // Bad: We just put i = 10 here and this was not caught by any test
+        // @todo: Open an issue? Write a new test?
         List<Pet> pets = new LinkedList<Pet>();
         Pet pet;
 
-        do {
-            pet = new Pet();
+        String [] lines = csv.split("\n");
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
+        for(String line: lines){
+            String [] fields = line.split(";");
+            // Process one line of the csv file
+            pet = new Pet();
 
-            pet.setName(field);
+            boolean success = false;
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
+            assert(fields.length >= 4);
+            assert(fields.length <= 5);
 
-            try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
+            setName(pet, fields[0]);
+            success = setBirthDate(pet, fields[1]);
+            if(!success){
                 HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
+                headers.add("errors", "date " + fields[1] + " not valid");
                 return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
             }
```

Comment:
We now generally handle errors on top-level. On low-level we just return a boolean whether the operation was successful. Still todo: Creating the error message is unclean.